Tabindex edits (#22882)
mdn/content: files/en-us/web/html/global_attributes/tabindex/index.md

{{HTMLSidebar("Global_attributes")}}

The **`tabindex`** [global attribute](/en-US/docs/Web/HTML/Global_attributes) allows developers to make html-elements focusable, allow or prevent them from being sequentially focusable (usually with the <kbd>Tab</kbd> key, hence the name), and determine their relative ordering for sequential focus navigation.

{{EmbedInteractiveExample("pages/tabbed/attribute-tabindex.html","tabbed-standard")}}

It accepts an integer as a value, with different results depending on the integer's value:

- A _negative value_ (the exact negative value doesn't actually matter, usually `tabindex="-1"`) means that the element is not reachable via sequential keyboard navigation, but could be focused with JavaScript (by calling the [`focus()`](/en-US/docs/Web/API/HTMLElement/focus) method) or visually by clicking with the mouse.

>   > **Note:** `tabindex="-1"` may be useful for elements that should not be navigated to directly using the <kbd>Tab</kbd> key, but need to have keyboard focus set to them. Examples include a modal dialog window that should be focused when it is opened, or a form submission error message that should be immediately focused when an errant form is submitted.

Reviewer: drive by comment: tabindex should never be set on a dialog. https://developer.mozilla.org/en-US/docs/Web/HTML/Element/dialog#attributes

Author: Your are absolutly right. The tabindex attribute must not be used on the <dialog> element. I will add it.
https://html.spec.whatwg.org/multipage/interactive-elements.html#the-dialog-element

Here I mean the case when we have a hidden (off-screen) dialog window for instance, login window or, (not exactly html-element <dialog>) which is shouldn't be reachable with keyboard (Tab key) but there are <input> elements in this dialog window which should be focused with JavaScript on some event (on button click or form submission).

`modal dialog window` sounds confusing. I can offer:

```suggestion
  > **Note:** `tabindex="-1"` may be useful for elements that should not be navigated to directly using the <kbd>Tab</kbd> key, but need to have keyboard focus set to them. Examples include a off-screen modal window that should be focused when it is opened, or a form submission error message that should be immediately focused when an errant form is submitted.
```

{{HTMLSidebar("Global_attributes")}}

> The **`tabindex`** [global attribute](/en-US/docs/Web/HTML/Global_attributes) allows developers to make html-elements focusable, allow or prevent them from being sequentially focusable (usually with the <kbd>Tab</kbd> key, hence the name), and determine their relative ordering for sequential focus navigation.

Reviewer 2: I find "make HTML elements focusable, allow or prevent them from being sequentially focusable" slightly repetitive. There's probably some overlapping semantics here but I don't have good wordings to suggest, and I'm fine to only make the capitalization change.

{{EmbedInteractiveExample("pages/tabbed/attribute-tabindex.html","tabbed-standard")}}

It accepts an integer as a value, with different results depending on the integer's value:

- A _negative value_ (the exact negative value doesn't actually matter, usually `tabindex="-1"`) means that the element is not reachable via sequential keyboard navigation, but could be focused with JavaScript (by calling the [`focus()`](/en-US/docs/Web/API/HTMLElement/focus) method) or visually by clicking with the mouse.

  > **Note:** `tabindex="-1"` may be useful for elements that should not be navigated to directly using the <kbd>Tab</kbd> key, but need to have keyboard focus set to them. Examples include a modal dialog window that should be focused when it is opened, or a form submission error message that should be immediately focused when an errant form is submitted.

> - `tabindex="0"` means that the element should be focusable in sequential keyboard navigation, after any positive tabindex values and its order is defined by the document's source order. The element could be focused with JavaScript (by calling the [`focus()`](/en-US/docs/Web/API/HTMLElement/focus) method) or visually by clicking with the mouse.

Author: You are right, it is duplicated from above. You got my point. `0` and `-1` are the most used as `tabindex` values and I was going to emphasizes that elements with `tabindex=”0”` also have programmatic focus and can be focused by users with mouse.

You are right, `tabindex` doesn’t fully cover topic `Focusable elements in HTML` because there are more cases when an element can become focusable (browsers have a list of elements which are focusable by default even without specified tablindex attribute). Tabindex value can be overridden by CSS (if an element or its parents have display: none or visibility:hidden) or `hidden` attribute and these elements will not be focusable and tabindex value really doesn't matter. For example, element with `display: none` and `tabindex=5` will not be focusable. And may be all these is out `tabindex` if scope.

But `tabindex` is relevant to the topic ` focusable elements in HTML` because developer can make an element focusable by adding ` tabindex` attribute with valid value.

To summarize, if an HTML element has `tabindex` attribute with any of valid value (for example, -3, -1, 0, 1, 5) this element can be focused with JavaScript (by developers) or visually by clicking by users with the mouse (or tapping on mobile) for every `tabindex` value.

And agree with your offer and I can remove this point from each item in the list and will explicitly specify this point before or after the list. Let me know if I need to do this.

- A _positive value_ means the element should be focusable in sequential keyboard navigation, with its order defined by the value of the number. That is, `tabindex="4"` is focused before `tabindex="5"` and `tabindex="0"`, but after `tabindex="3"`. If multiple elements share the same positive `tabindex` value, their order relative to each other follows their position in the document source. The maximum value for `tabindex` is 32767.
- If the `tabindex` value is omitted the user agent will decide whether the element is focusable.

>   > **Warning:** It is recommended that developers use only 0 and -1 as `tabindex` values. Avoid using `tabindex` values greater than 0 and css-properties that can change the order of focusable html-elements ([Ordering flex items](/en-US/docs/Web/CSS/CSS_Flexible_Box_Layout/Ordering_Flex_Items)). Doing so makes it difficult for people who rely on using keyboard for navigation or assistive technology to navigate and operate page content. Instead, write the document with the elements in a logical sequence.

Reviewer 2: ```suggestion
  > **Warning:** You are recommended to only use `0` and `-1` as `tabindex` values. Avoid using `tabindex` values greater than `0` and CSS properties that can change the order of focusable HTML elements ([Ordering flex items](/en-US/docs/Web/CSS/CSS_Flexible_Box_Layout/Ordering_Flex_Items)). Doing so makes it difficult for people who rely on using keyboard for navigation or assistive technology to navigate and operate page content. Instead, write the document with the elements in a logical sequence.
```

> Some focusable html-elements are determined (by [user agent](/en-US/docs/Glossary/User_agent) under hood) to have `tabindex` default value equals 0. These elements are an {{HTMLElement("a")}} and {{HTMLElement("area")}} with `href` attribute, {{HTMLElement("button")}}, {{HTMLElement("frame")}} {{deprecated_inline}}, {{HTMLElement("iframe")}}, {{HTMLElement("input")}}, {{HTMLElement("object")}}, {{HTMLElement("select")}}, {{HTMLElement("textarea")}}, or SVG {{SVGElement("a")}} element, or is a {{HTMLElement("summary")}} element that is a summary for its parent {{HTMLElement("details")}}. Developers shouldn't add the `tabindex` attribute to these elements only if they are going to change the default behavior.

Reviewer 2: ```suggestion
Some focusable HTML elements have a default `tabindex` value of 0 set under the hood by the [user agent](/en-US/docs/Glossary/User_agent). These elements are an {{HTMLElement("a")}} or {{HTMLElement("area")}} with `href` attribute, {{HTMLElement("button")}}, {{HTMLElement("frame")}} {{deprecated_inline}}, {{HTMLElement("iframe")}}, {{HTMLElement("input")}}, {{HTMLElement("object")}}, {{HTMLElement("select")}}, {{HTMLElement("textarea")}}, or SVG {{SVGElement("a")}} element, or a {{HTMLElement("summary")}} element that provides summary for a {{HTMLElement("details")}} element. Developers shouldn't add the `tabindex` attribute to these elements unless it changes the default behavior.
```

Makes sense, but it should be "under the hood" in that case 😄

> Some focusable html-elements are determined (by [user agent](/en-US/docs/Glossary/User_agent) under hood) to have `tabindex` default value equals 0. These elements are an {{HTMLElement("a")}} and {{HTMLElement("area")}} with `href` attribute, {{HTMLElement("button")}}, {{HTMLElement("frame")}} {{deprecated_inline}}, {{HTMLElement("iframe")}}, {{HTMLElement("input")}}, {{HTMLElement("object")}}, {{HTMLElement("select")}}, {{HTMLElement("textarea")}}, or SVG {{SVGElement("a")}} element, or is a {{HTMLElement("summary")}} element that is a summary for its parent {{HTMLElement("details")}}. Developers shouldn't add the `tabindex` attribute to these elements only if they are going to change the default behavior.

Author: Oops, non-native speaker is detected 😄. 

Sorry, I have some problems with translation. Thanks for your quick answers and fixes.

{{HTMLSidebar("Global_attributes")}}

The **`tabindex`** [global attribute](/en-US/docs/Web/HTML/Global_attributes) allows developers to make HTML elements focusable, allow or prevent them from being sequentially focusable (usually with the <kbd>Tab</kbd> key, hence the name) and determine their relative ordering for sequential focus navigation.

{{EmbedInteractiveExample("pages/tabbed/attribute-tabindex.html","tabbed-standard")}}

It accepts an integer as a value, with different results depending on the integer's value:

>   > **Note:** If an HTML element renders and has `tabindex` attribute with any valid integer's value, so this element can be focused with JavaScript (by calling the [`focus()`](/en-US/docs/Web/API/HTMLElement/focus) method) or visually by clicking with the mouse.

Reviewer: ```suggestion
  > **Note:** If an HTML element renders and has `tabindex` attribute with any valid integer value, the element can be focused with JavaScript (by calling the [`focus()`](/en-US/docs/Web/API/HTMLElement/focus) method) or visually by clicking with the mouse. If the value is `0` or greater, the element can also be focused via the tab key.
```

Author: We discussed this point with @Josh-Cena - https://github.com/mdn/content/pull/22882#discussion_r1047734948. 

The idea was to specifiy common behavior of HTML-elements with `tabindex` attribute before the list to avoid duplicates in each item of this list.

Please, consider the following:

```suggestion
> **Note:** If an HTML element renders and has `tabindex` attribute with any valid integer value, the element can be focused with JavaScript (by calling the [`focus()`](/en-US/docs/Web/API/HTMLElement/focus) method) or visually by clicking with the mouse. The particular `tabindex` value controls whether the element is `tabbable` (i.e. reachable via sequential keyboard navigation, usually with the <kbd>Tab</kbd> key).
```

- A _negative value_ (the exact negative value doesn't actually matter, usually `tabindex="-1"`) means that the element is not reachable via sequential keyboard navigation.

>   > **Note:** `tabindex="-1"` may be useful for elements that should not be navigated to directly using the <kbd>Tab</kbd> key, but need to have keyboard focus set to them. Examples include a off-screen modal window that should be focused when it is opened, or a form submission error message that should be immediately focused when an errant form is submitted.

Reviewer: ```suggestion
  > **Note:** `tabindex="-1"` may be useful for elements that should not be navigated to directly using the <kbd>Tab</kbd> key, but need to have keyboard focus set to them. Examples include an off-screen modal window that should be focused when it comes into view, or a form submission error message that should be immediately focused when an errant form is submitted.
```

> - `tabindex="0"` means that the element should be focusable in sequential keyboard navigation, after any positive tabindex values. Focus navigation order of these elements is defined by their order in the document source.

Reviewer: ```suggestion
- `tabindex="0"` means that the element should be focusable in sequential keyboard navigation, after any positive `tabindex` values. The focus navigation order of these elements is defined by their order in the document source.
```

> - `tabindex="0"` means that the element should be focusable in sequential keyboard navigation, after any positive tabindex values. Focus navigation order of these elements is defined by their order in the document's source.

Reviewer: ```suggestion
- `tabindex="0"` means that the element should be focusable in sequential keyboard navigation, after any positive tabindex values. Focus navigation order of these elements is defined by their order in the document source.
```

- A _positive value_ means the element should be focusable in sequential keyboard navigation, with its order defined by the value of the number. That is, `tabindex="4"` is focused before `tabindex="5"` and `tabindex="0"`, but after `tabindex="3"`. If multiple elements share the same positive `tabindex` value, their order relative to each other follows their position in the document source. The maximum value for `tabindex` is 32767.

> - If the `tabindex` value is omitted the user agent will decide whether the element is focusable.

Reviewer: ```suggestion
- If the `tabindex` attribute is included with no value set, whether the element is focusable is determined by the user-agent.
```

Author: user-agent => user agent
```suggestion
- If the `tabindex` attribute is included with no value set, whether the element is focusable is determined by the user agent.
```

  > **Warning:** You are recommended to only use `0` and `-1` as `tabindex` values. Avoid using `tabindex` values greater than `0` and CSS properties that can change the order of focusable HTML elements ([Ordering flex items](/en-US/docs/Web/CSS/CSS_Flexible_Box_Layout/Ordering_Flex_Items)). Doing so makes it difficult for people who rely on using keyboard for navigation or assistive technology to navigate and operate page content. Instead, write the document with the elements in a logical sequence.

> Some focusable HTML elements have a default `tabindex` value of 0 set under the hood by the [user agent](/en-US/docs/Glossary/User_agent). These elements are an {{HTMLElement("a")}} or {{HTMLElement("area")}} with `href` attribute, {{HTMLElement("button")}}, {{HTMLElement("frame")}} {{deprecated_inline}}, {{HTMLElement("iframe")}}, {{HTMLElement("input")}}, {{HTMLElement("object")}}, {{HTMLElement("select")}}, {{HTMLElement("textarea")}}, or SVG {{SVGElement("a")}} element, or a {{HTMLElement("summary")}} element that provides summary for a {{HTMLElement("details")}} element. Developers shouldn't add the `tabindex` attribute to these elements unless it changes the default behavior.

Reviewer: ```suggestion
Some focusable HTML elements have a default `tabindex` value of `0` set under the hood by the [user agent](/en-US/docs/Glossary/User_agent). These elements are an {{HTMLElement("a")}} or {{HTMLElement("area")}} with `href` attribute, {{HTMLElement("button")}}, {{HTMLElement("frame")}} {{deprecated_inline}}, {{HTMLElement("iframe")}}, {{HTMLElement("input")}}, {{HTMLElement("object")}}, {{HTMLElement("select")}}, {{HTMLElement("textarea")}}, or SVG {{SVGElement("a")}} element, or a {{HTMLElement("summary")}} element that provides summary for a {{HTMLElement("details")}} element. Developers shouldn't add the `tabindex` attribute to these elements unless it changes the default behavior.
```

Author: ```suggestion
Some focusable HTML elements have a default `tabindex` value of `0` set under the hood by the [user agent](/en-US/docs/Glossary/User_agent). These elements are an {{HTMLElement("a")}} or {{HTMLElement("area")}} with `href` attribute, {{HTMLElement("button")}}, {{HTMLElement("frame")}} {{deprecated_inline}}, {{HTMLElement("iframe")}}, {{HTMLElement("input")}}, {{HTMLElement("object")}}, {{HTMLElement("select")}}, {{HTMLElement("textarea")}}, and SVG {{SVGElement("a")}} element, or a {{HTMLElement("summary")}} element that provides summary for a {{HTMLElement("details")}} element. Developers shouldn't add the `tabindex` attribute to these elements unless it changes the default behavior (for instance, including a negative value will remove the element from the focus navigation order).
```

> Some focusable HTML elements have a default `tabindex` value of `0` set under the hood by the [user agent](/en-US/docs/Glossary/User_agent). These elements are an {{HTMLElement("a")}} or {{HTMLElement("area")}} with `href` attribute, {{HTMLElement("button")}}, {{HTMLElement("frame")}} {{deprecated_inline}}, {{HTMLElement("iframe")}}, {{HTMLElement("input")}}, {{HTMLElement("object")}}, {{HTMLElement("select")}}, {{HTMLElement("textarea")}}, and SVG {{SVGElement("a")}} element, or a {{HTMLElement("summary")}} element that provides summary for a {{HTMLElement("details")}} element. Developers shouldn't add the `tabindex` attribute to these elements unless it changes the default behavior (for instance, including a negative value will remove the element from the focus navigation order).

Reviewer: ```suggestion
Some focusable HTML elements have a default `tabindex` value of `0` set under the hood by the [user agent](/en-US/docs/Glossary/User_agent). These elements are an {{HTMLElement("a")}} or {{HTMLElement("area")}} with `href` attribute, {{HTMLElement("button")}}, {{HTMLElement("frame")}} {{deprecated_inline}}, {{HTMLElement("iframe")}}, {{HTMLElement("input")}}, {{HTMLElement("object")}}, {{HTMLElement("select")}}, {{HTMLElement("textarea")}}, and SVG {{SVGElement("a")}} element, or a {{HTMLElement("summary")}} element that provides summary for a {{HTMLElement("details")}} element. Developers shouldn't add the `tabindex` attribute to these elements unless it changes the default behavior (for example, including a negative value will remove the element from the focus navigation order).
```
